docs/src/app/components/examples/app-bar-waterfall-example.jsx

```diff
@@ -0,0 +1,79 @@
+import React from 'react';
+import {AppBar} from 'material-ui';
+
+import IconButton from 'icon-button';
+import MoreVertIcon from 'svg-icons/navigation/more-vert';
+import ArrowBack from 'svg-icons/navigation/arrow-back';
+
+const AppBarWaterfallExample = React.createClass({
+
+  render() {
+    return (
+      <AppBar
+        position="waterfall"
+        waterfall={{
+          minHeight: 64,
+          maxHeight: 210,
+          children: this.getWaterfallChildren(),
+          onHeightChange: this.onHeightChange,
+        }}
+        title={
+          <div
+            ref={el => { this.titleEl = el; }}>
+            Waterfall AppBar
+          </div>
+        }
+        iconElementLeft={
+          <IconButton onClick={this.onBackClick}>
+            <ArrowBack />
+          </IconButton>
+        }
+        iconElementRight={
+          <IconButton>
+            <MoreVertIcon/>
+          </IconButton>
+        }
+        />
+    );
+  },
+
+  getWaterfallChildren() {
+    let styles = this.getStyles();
+    return (
+      <div
+        style={{overflow: 'hidden'}}>
+        <img
+          ref={el => { this.logoEl = el; }}
+          style={styles.logo}
+          src="images/material-ui-logo.svg"/>
+      </div>
+    );
+  },
+
+  onHeightChange({height, minHeight, maxHeight}) {
+    let interpolation = (height - minHeight) / (maxHeight - minHeight);
+
+    // For best performance, we will directly modify style on DOM elements
```

Comment:
I needed this type of optimization for mobile version. But I will modify the example to add also the simple, un-optimized version, that just use setState and re-renders.